### PR TITLE
Use new record equality specification 

### PR DESCRIPTION
**Pull request by agocke**
Per design changes, record equality is now defined using all instance
fields, instead of just record fields.

Spec changes at https://github.com/dotnet/csharplang/pull/3396

### CODE REVIEW

**Comment by agocke** [comment]
@gafter @333fred for review

File: src/Compilers/CSharp/Test/Symbol/Symbols/AnonymousTypesSymbolTests.cs [hunk 1]
```diff
@@ -0,0 +1,1911 @@
...
+                                               new TypeDescr() { FieldNames = new string[] { "x", "g" } },
+                                               new TypeDescr() { FieldNames = new string[] { "<>h__TransparentIdentifier0", "z" } }
+                                           )
+            );
+
+            TestAnonymousTypeFieldSymbols_InQuery(verifier.EmittedAssemblyData);
+        }
+
+        [ClrOnlyFact]
+        public void AnonymousTypeSymbol_Mix()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/AnonymousTypesSymbolTests.cs:56]: What is the purpose of this test?

File: src/Compilers/CSharp/Test/Symbol/Symbols/AnonymousTypesSymbolTests.cs [hunk 2]
```diff
@@ -0,0 +1,1911 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Emit;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/AnonymousTypesSymbolTests.cs:21]: I'm generally unsure what the addition of these tests are signifying in this PR. Was this a move?

**Comment by agocke** [comment]
ping @gafter for review

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs [hunk 3]
```diff
@@ -144,25 +142,25 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
                 retExpr = F.ObjectNotEqual(other, F.Null(F.SpecialType(SpecialType.System_Object)));
             }
 
-            var recordProperties = ArrayBuilder<FieldSymbol>.GetInstance();
+            var fields = ArrayBuilder<FieldSymbol>.GetInstance();
             foreach (var member in ContainingType.GetMembers())
             {
                 // PROTOTYPE: Should generate equality on user-written members as well
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:148]: >PROTOTYPE [](start = 19, length = 9)

This comment doesn't make sense.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs [hunk 4]
```diff
@@ -144,25 +142,25 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
                 retExpr = F.ObjectNotEqual(other, F.Null(F.SpecialType(SpecialType.System_Object)));
             }
 
-            var recordProperties = ArrayBuilder<FieldSymbol>.GetInstance();
+            var fields = ArrayBuilder<FieldSymbol>.GetInstance();
             foreach (var member in ContainingType.GetMembers())
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:146]: >GetMembers [](start = 50, length = 10)

This does not include fieldlike event backing fields.

File: src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.Lowered.cs [hunk 5]
```diff
@@ -253,6 +251,61 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
...
+                    F.Field(F.This(), field),
+                    F.Field(otherReceiver, field));
+
+                // Generate 'retExpression' = 'retExpression && nextEquals'
+                retExpression = retExpression is null
+                    ? nextEquals
+                    : F.LogicalAnd(retExpression, nextEquals);
+            }
+
+            RoslynDebug.AssertNotNull(retExpression);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.Lowered.cs:304]: >AssertNotNull [](start = 24, length = 13)

This assertion does not appear justified.

**Comment by gafter** [comment]
Finished a quick scan of Iteration 2.

File: src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.Lowered.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.Lowered.cs:304]: Earlier `fields` is asserted to have more than 1 element, there will always be at least one iteration of the `foreach`.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs (hunk 3 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs]: Outdated, removed

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs (hunk 4 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs]: Good catch. Updated and test added.

File: src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.Lowered.cs (hunk 5 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.Lowered.cs:304]: How is there always at least one iteration of the foreach?  According to the spec a record can have an empty parameter list.

---
In reply to: [418268985](https://github.com/dotnet/roslyn/pull/43626#discussion_r418268985) [](ancestors = 418268985)

File: src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.Lowered.cs [hunk 6]
```diff
@@ -253,6 +251,61 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
...
+        /// `this` are equal to the fields on <paramref name="otherReceiver" /> according to the
+        /// default EqualityComparer.
+        /// </summary>
+        public static BoundExpression GenerateFieldEquals<TList>(
+            BoundExpression? initialExpression,
+            BoundExpression otherReceiver,
+            TList fields,
+            SyntheticBoundNodeFactory F) where TList : IReadOnlyList<FieldSymbol>
+        {
+            Debug.Assert(fields.Count > 0);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.Lowered.cs:265]: >fields.Count > 0 [](start = 25, length = 16)

How is this justified?

**Pull request merged**